docs/design_spec.md

```diff
@@ -0,0 +1,94 @@
+# Nimbess Agent Design
+
+## Overview
+
+Nimbess Agent's main purpose is to manage an external data plane. The Agent's main responsibilities
+include:
+
+ 1. Translating Northbound (NB) application-specific resources into network oriented resources across the
+ entire network fabric.
+ 2. Managing and serving as the control plane for the locally assigned, and external data plane (Southbound).
+ 
+ The Agent is distributed over every host and each Agent instance is capable of viewing and managing the
+ network configuration for the entire fabric. However, only the Agent local to a data plane (on the same host)
+ is capable of directly configuring that data plane.
+ 
+## Agent Northbound Design
+ 
+Nimbess Agent is able to support multiple container platforms and by default supports Kubernetes (K8S). It
+also supports the CNI spec for handling CNI requests from Kubelet or other container platforms which use
+CNI. The Nimbess Agent NB is responsible for translating K8S resources into network-oriented resources.
+
+The Agent exposes a gRPC based CNI API in order to service CNI requests. The requests are translated into
+Nimbess Ports, and a Network Pipeline is built for that Port. For other networking related container platform
+resources, such as a K8S Service, Ingress, etc; Nimbess Agent will handle translating those resources into their
+relevant network-oriented constructs and applied to one or more pipelines.
+ 
+## Network Constructs in Nimbess Agent
+ 
+In Nimbess, how traffic flows through a data plane can be though of as a pipeline of Network Functions.
+Traffic ingresses a data plane, flows through a series of network functions such as a Switch, ACL,
+Load-Balancer, etc and then egresses the data plane out of a port. Nimbess uses this concept of a Network
+Pipeline and creates one for each Port that is attached to the data plane. There are multiple reasons for
+this design:
+ 
+ 1. Guarantees loop-free forwarding
+ 2. Increases per network function processing efficiency
+ 3. Simplifies debugging
+ 
+When a Port is added to a pipeline, it is broken up into an IngressPort as well as an EgressPort. Only the
+IngressPort is ever attached to this Port's pipeline, while the EgressPort is attached to every other pipeline which
+belongs to the same network fabric (L2/L3 domain). This prevents traffic from ever being forwarded back
+out the same interface from which it originated (1). Additionally, this design increases network
+function processing efficiency, because the cache of the tables are limited to lookups for traffic that may ingress
+a single port. Debugging becomes easier because the only traffic in a particular pipeline only ever comes
+from a single port. However, one of the drawbacks to this design is that there is increased memory usage,
+as the number of instances of network functions scale by the number of ports in the system.
+
+Although each Port contains its own Port Pipeline, the core network functions portion of the pipeline
```

Comment:
is a port pipeline a real construct? or it is virtual - each one is the same virtual construct? Or does each port pipeline really a unique construct built up of various functions; each port comes up and the pipeline is built for that port, rather that it is given a predetermined virtual pipeline

Reply:
Each port has its own pipeline built of independent module instances. The core part of the pipeline for the port is decided by the meta pipeline. The meta pipeline modules are copied and used to create a new port pipeline. The Ingress and Egress Port modules are created from scratch.